### PR TITLE
WIP: Add RStudio

### LINKED ISSUE #13424: Package RStudio / RStudio server
Anaconda [has some](https://anaconda.org/r/rstudio) heavily outdated versions in the `R` channel still, but I cannot find the recipe in [AnacondaRecipes/aggregate](https://github.com/AnacondaRecipes/aggregate/search?q=rstudio).

The [license](https://github.com/rstudio/rstudio/blob/master/COPYING) would be compatible.

I'm guessing the topic has come up before (not least since solving integration with R / RStudio / conda is a recurring subject), so I wanted to open this issue mainly for getting some thoughts from @conda-forge/core on this.

If there are no strong objections, I'll probably try my hand at a PR in the coming days.

### PR DESCRIPTION
**Pull request by h-vetinari**
Closes #13424, where the licensing question was discussed, and an explicit OK from RStudio was [provided](https://github.com/conda-forge/staged-recipes/issues/13424#issuecomment-742830241).

@mbargull [pointed](https://github.com/conda-forge/staged-recipes/issues/13424#issuecomment-740840236) me to @mingwandroid's work on this in [aggregateR](https://github.com/AnacondaRecipes/aggregateR/tree/master/rstudio-feedstock/recipe), which I've taken over with the least amount of changes to satisfy the linter, plus a few small clean-ups.

I wouldn't be surprised if this takes several iterations (month?) to get right, so _any_ help is welcome!

TODO:
* [x] unblock environment resolution - unix
* [ ] unblock environment resolution - win (this [worked](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=268210&view=logs&jobId=6f142865-96c3-535c-b7ea-873d86b887bd&j=240f1fee-52bc-5498-a14a-8361bde76ba0&t=23010c01-6a9d-5f94-9037-792618644b77) for https://github.com/conda-forge/staged-recipes/pull/13760/commits/3e64cc17de8e7f65c803d45a070dcd3ac2f12980 but then failed with the same setup in https://github.com/conda-forge/staged-recipes/pull/13760/commits/19ecbc5150b7eb3b23d02391aae1e1a316422030)
* [x] unblock patch application 
* [x] depend on newer `r-base`j
* [ ] get package building
* [ ] add actual tests
* [ ] turn into multi-output recipe
* [ ] add output for rstudio-server
* [ ] remove dependencies coming from defaults (?)
* [ ] adapt default `reticulate` setup (?)
* [ ] avoid stuff from https://s3.amazonaws.com/rstudio-buildtools where possible (?)
* [ ] add more maintainers

### CODE REVIEW

**Comment by h-vetinari** [comment]
First hurdle, `soci`:
```
conda.exceptions.ResolvePackageNotFound: 
  - soci
```

Seems that Ray's work depends on unpublished packages, because neither does `main` contain `soci` nor `qt 5.15`. Let's see if we can replace with the cf versions...

File: recipes/rstudio/meta.yaml [hunk 1]
```diff
@@ -49,7 +49,6 @@ source:
       - patches/0039-Add-support-for-Qt-5.10.patch
       - patches/0040-Revert-disable-macOS-specific-Cmd-Shift-handling-clo.patch
       - patches/0041-Fix-missing-boost-placeholders.patch
-      - patches/0042-Conda-We-use-a-shared-soci-library.patch
```
Comment (h-vetinari) [inline recipes/rstudio/meta.yaml]: There was nothing about `SOCI_LIBRARIES` in 1.3.1073, particularly not in `src/cpp/CMakeLists.txt`. Skipped this patch.

**Comment by h-vetinari** [comment]
The rebase didn't apply cleanly to `1.3.1073`, had to rebase some patches manually. I suggest reviewing https://github.com/conda-forge/staged-recipes/pull/13760/commits/9c2d23d93c0296f5ab8098c4696ddf847a67e8a6 with whitespace changes ignored. Very likely a line-ending thing, which I can fix if absolutely necessary (but these are just the output of `git format-patch tags/v1.3.1073`; on windows, no less...)

**Comment by h-vetinari** [comment]
@thomann contacted my privately, added him as a contributor.

Glad to have more people participate! 🙃

**Comment by dhirschfeld** [comment]
Thanks for this PR @h-vetinari! It would be great to have `rtstudio` available from `conda-forge`.

What's the current status of this PR? Are there any blockers? Anything others can help with?

**Comment by h-vetinari** [comment]
> What's the current status of this PR? Are there any blockers? Anything others can help with?

Hey @dhirschfeld - thanks for checking in! I'm currently pretty swamped with other things, and cracking this nut will take some dedicated time I don't foresee to have soon. So I'm happy to give push access to this PR to anyone who wants to contribute.

As another substantial hurdle that arose since opening that PR, Ray Donnelly [tragically](https://twitter.com/ocefpaf/status/1384619253643550732) died, so understanding the recipe he created almost single-handedly (much less his patches) likely got substantially harder.

**Comment by dhirschfeld** [comment]
Thanks for the update @h-vetinari. I won't have time myself in the near future either so will leave this on the back-burner for now...

Ray's passing was a great loss to the community - a lot of us benefited from his sharing of his time and knowledge 😢

**Comment by thomann** [comment]
Hey @h-vetinari!

So I'm a little lost with the multiple-output part here - the build-requirements are not available in the build-scripts. I tried a couple of things in the last months but never got an inch ahead. If no one else has an insight I'd propose to do separate recipes for Server and for Desktop.

As a minimal example take the following meta.yaml:
```
package:
  name: test-output
  version: 1.4.1103
source:
  - url: https://github.com/rstudio/rstudio/archive/v1.4.1103.tar.gz
    sha256: e448aaaf7ac7f4fd97197250762bfd28195c71abfd67db6f952463dea552be4c
build:
  number: 0
outputs:
  - name: rstudio-desktop
    script: build.sh  # [unix]
    requirements:
      build:
        - make
      run:
        - openjdk
  - name: rstudio-server
    script: build-server.sh  # [unix]
    requirements:
      build:
        - make
      run:
        - jq
```
And say the build.sh:
```
#!/bin/bash
which make
```
Then that fails because make is not available.

**Comment by thomann** [comment]
OK, I figured out the problem with the outputs build: both in the example and in the PR we have a build.sh and a build-server.sh. However, the build.sh triggers that conda-build thinks to use that already in the "root" requirements of the meta.yaml etc. But then we cannot have different build scripts for the outputs. So I renamed it to build-desktop.sh and now it does as expected.

**Comment by h-vetinari** [comment]
Hey @thomann, sorry for the delayed response. Great that you're picking this up again - I see you already figured out the build-script part (`build.sh` / `bld.bat` always get executed by conda-build by default; in the case of a multi-output recipe, these get run as a global build-step before any of the outputs [sometimes some build steps can be shared, or indeed need to be]).

I'm not super familiar with the cdt-infrastructure myself - it's great that you've already opened a PR, let's see where this leads. Though maybe it would be enough for the moment to simply add them as [yum-requirements](https://conda-forge.org/docs/maintainer/knowledge_base.html#yum-deps)?

**Pull request closed without merging**